Aqui empieza el curso de ML aplicado

In [7]:
import datetime
print(f'The last time is:  {datetime.datetime.now()}')

The last time is:  2026-04-06 21:56:20.247632


In [8]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import sklearn

Bueno, lo mas importante a la hora de estudiar ML y los modelos es entender los pasos de lo que estamos haciendo, entender la teoria, identificar el problema, tener los datos libros. 
Si tenemos un problema de clasificacion, un problema de prediccion o un problema de agrupacion. Los pasos que tenemos que considerar para el workflow de Scikit-learn son: 
1. Getting Data (find columns, labels, prepair train and test steps)
2. Choosing a model for our problem 
3. Ajustar el modelo a la data que tenemos y usarlo para obtener la predicción
4. Evaluar el modelo 
5. Experimentar para mejorar
6. Guardar el modelo para que alguien mas lo pueda usar, o yo mismo en el futuro. 

Para el primer ejemplo vamos a usar un dataset llamado heart-disease.csv

In [9]:
df = pd.read_csv("https://raw.githubusercontent.com/Juan-Vega18/ML_full_course/refs/heads/main/my_notebooks/data/heart-disease.csv?token=GHSAT0AAAAAADWIPU2PWGOYYCMMKIMUB6YM2OUQBNA")
df

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0


In [10]:
#Ahora vamos a ver algunas propiedades de nuestro dataset
df.info()
df.shape
df.head(5)
#df['target'].unique()
#df['restecg'].unique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


Encontramos un conjunto de 14 columnas con 303 filas. La fase de entender como y que especificamente compone nuestro dataset es absolutamente clave. Tenemos target que a pesar de que parece ser una variable numerica en realidad es booleana, el dataset esta completisimo, no tiene vacios etc. 

En el casode este proyecto, esa variable booleana es precisamente la que queremos predecir, si está o no enfermo del corazón, 1 significa si, 0 significa no. Las demas columnas se llamaran 'features' y las usaremos para intentar predecir a target. 

In [11]:
x = df.drop('target', axis=1) #Creamos los x de los parametros. eliminamos la columna target
y = df['target'] #Creamos la y que será nuestra predicción 
x.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2


In [12]:
#Ahora vamos a revisar los valores de y que tenemos 
print(f'Los valores de la variable estan distribuidos de la siguiente forma: {y.value_counts()}')
y.unique()

Los valores de la variable estan distribuidos de la siguiente forma: target
1    165
0    138
Name: count, dtype: int64


array([1, 0])

Ahora, algo que es absolutamente importante en el ML es entender que la data debe ser divida en dos grupos, training set and test set. Es decir, un grupo de datos se usará para entrenar el modelo, el otro para probarlo. Usualmente se combina en proporciones como 80/20, 70/30, 75/25, tambien es posible tener el validation set que es un dataset extra que se usa durante el entrenamiento pero por ahora no lo vamos a usar. 

Scikit-learn ya incluye sklearn.model_selection.train_test_split que es la funcion de la libreria para dividir el dataset en training sets. 


In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.25)
print(X_train.shape)
print(type(X_train))
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

#Aqui ya dividimos la data, notemos que la funcion train_test_split asigna a 4 variables los 
#datos correspondientes y los subdivide en pandas.core.frame.DataFrame luego veremos que significa
# y que tiene de diferente con el pd.Dataframe tipico.

X_train

(227, 13)
<class 'pandas.core.frame.DataFrame'>
(76, 13)
(227,)
(76,)


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
16,58,0,2,120,340,0,1,172,0,0.0,2,0,2
132,42,1,1,120,295,0,1,162,0,0.0,2,0,2
8,52,1,2,172,199,1,1,162,0,0.5,2,0,3
179,57,1,0,150,276,0,0,112,1,0.6,1,1,1
254,59,1,3,160,273,0,0,125,0,0.0,2,0,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...
243,57,1,0,152,274,0,1,88,1,1.2,1,1,3
64,58,1,2,140,211,1,0,165,0,0.0,2,0,2
20,59,1,0,135,234,0,1,161,0,0.5,1,0,3
296,63,0,0,124,197,0,1,136,1,0.0,1,0,2


Ahora que tenemos los datos divididos es necesario escoger el modelo y los hiperparametros. Escoger el modelo depende normalmente de cual sea el problema que estamos trabajando, por ejemplo, Scikit-learn tiene recomendaciones si el problema es de clasificacion o predicción. 
En la web de scikit-learn tenemos un mapa muy util que explica que modelo usar, cuando y bajo que condiciones. 
La primera y mas importante es la cantidad de ejemplos (se recomienda mas de 50). 

Tenemos problemas de clasificacion, regresion, clustering y reduccion de la dimension. 
Aquí es donde el estudiante (yo) debería aprender bien y antes de que el curso acabe haber probado todos o la gran mayoría de los algoritmos, la programación no necesariamente es lo mas importante, lo que importa aquí es saber bien que modelo usar, cuando, y por qué. 

Tambien tenemos los hiperparametros del modelo, que si sabemos matematicamente lo que estamos haciendo, podemos tocarlos y revisar si el modelo mejora o no. Se van ajustando. Aquí tambien es muy importante que el estudiante entienda que está pasando en el interior de la maquina, de nada sirve tener el modelo, el codigo, resultados si no soy capaz de modificar parametros (no entiendo nada).

En el caso de este proyecto, estamos trabajando con un clasificador, booleano. Si o no. Asi que el curso nos recomienda usar empezar con el classifier Random forest.

In [14]:
from sklearn.ensemble import RandomForestClassifier #Aqui hay muchos interesantes 

clf = RandomForestClassifier() #Aqui inicializamos el modelo, y vamos a ver los parametros
#que podemos modificar 
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

#Por ahora, 
no vamos a tocarlos, los dejaremos para mas adelante. Usaremos entonces
#los parametros por defecto de Sckit-learn. 

Ahora, vamos a Fit the model to the data and use it to make a prediction. Es lo que queremos. Si tenemos etiquetas (aprendizaje supervisado) el modelo trabaja en encontrar la relacion entre los datos y las etiquetas. En el caso de que no hayan etiquetas, el modelo intenta encontrar patrones y agrupar ejemplos similares juntos. 
La mayoria de los modelos de Scikit-Learn tienen el methodo fit(X,Y) integrado, donde X son las caracteristicas (features) e y son las etiquetas. 
En nuestro caso, vamos a empezar ajustando un modelo a los datos de entrenamiento separados. 



In [ ]:

clf.fit(X=X_train, y=y_train) #clf es nuestro modelo, y a ese modelo aplicamos parametros

RandomForestClassifier()

Ahora, ya que tenemos el "modelo ajustado" viene la parte central de esto. intentar hacer una predicción. Despues de que el modelo ya está entrenado, ahora lo que hay que hacer es predecir, usando predict(). En otras palabras, usar el modelo con data nueva, que no haya visto y que no esté etiquetada. (Ojo, aqui es importente), la data que daremos para predecir debe tener el mismo shape que la data que usamos para entrenar (matematicamente tiene todo el sentido pero igual se aclara). 

Para eso, usaremos la data de test

In [31]:
y_preds = clf.predict(X=X_test)
print(y_preds)

#Y si queremos analizar una unica fila, dado que es un 
#dataset 2D esperamos esa entrada

predic_1 = clf.predict(X_test.iloc[[0]])
predic_1

[0 0 1 0 1 0 1 1 1 1 1 0 0 0 1 0 1 0 1 1 1 1 0 1 1 0 1 0 0 0 0 1 0 1 0 0 1
 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 0 0 0 1 0 0 1 0 1 1 0 0 0 0 0 1 0 1 0
 1 1]


array([0])

Ahí tenemos el array de predicciones, ahora lo que nos interesa es evaluar el modelo. Cada modelo estimador tiene un metodo integrado llamada score(). Este metodo compara que tan bien le fue al modelo aprendiendo los patrones entre las caracteristicas y las etiquetas. El metodo score() usa una metrica estandar de evaluacion para medir los resultados del modelo. 

En el caso de un clasificador (nuestro modelo), una de las metricas de evaluacion mas comunes es 'accuracy', que es la fraccion de predicciones correctas sobre el total de predicciones. 

In [38]:
#Evaluate the model on the training set

train_acc = clf.score(X=X_test, y=y_test)
print(f'The models accuracy on the training dataset is: {train_acc*100} %')

The models accuracy on the training dataset is: 85.52631578947368 %


Existen algunos otros metodos de evaluacion para nuestro modelo, estos vienen del modulo sklearn.metrics y tenemos: 
Classification report (y true, y true)
Construye un reporte mostrando varias metricas de clasificacion, como precision, recall, and F1-score. (luego las entenderemos)

Confusion matrix(y true, y pred). Crea una matriz de confusion para comparar la predición de las etiquetas verdaderas

Accuracy score (y_true, y_pred). Encuentra el puntaje de precisión, esta es la metrica por defecto de los clasificadores. 

In [40]:
#Classification report 
from sklearn.metrics import classification_report

print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.91      0.79      0.85        39
           1       0.81      0.92      0.86        37

    accuracy                           0.86        76
   macro avg       0.86      0.86      0.86        76
weighted avg       0.86      0.86      0.85        76



In [41]:
#Confusion matrix
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_preds))

[[31  8]
 [ 3 34]]


In [46]:
#Compute the accuracy score

from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_preds)

0.8552631578947368

Despues de todo esto que vimos, nos damos cuenta que en realidad la implementacion es muy sencilla, y obtuvimos de hecho un 'buen rendimiento', supongo, dado que en el futuro de este curso vamos a estudiar como saber si el rendimiento fue bueno o no etc. 

El primer modelo es usualmente tomado como baseline, basicamente lo usamos para revisar la predicción base y es un indicador de que debemos hacer en el futuro, aqui debe aparece un criterio importante y es ir aprendiendo cuando con el baseline vale la pena continuar con un estimador, o aprender que parametros parece mas importante modificar etc. Importante hacer la prueba e ir escalando. 

El siguiente paso es en el workflow es intentar mejorar nuestro modelo base (que en teoria ya deberia tener un margen decente para empezar)

Como? 
Bueno, una de las cosas mas importantes en el Machine Learning es experimentar, experimentar y experimentar, es la forma correcta. Y hay dos caminos que hoy podemos mencionar para los pasos a continuacion: 

1. Experimentos enfocados en el modelo
2. Experimentos enfocados en los datos

Los experimentos enfocados en el modelo pueden incluir modificar parametros mucho mas complejos o usar un modelo mas complicado, 

La perspectiva enfocada en los datos puede incluir obtener mas datos o de mejor calidad, eso podria hacer que el modelo que ya tenemos tenga mas chances de aprender mejor los patrones. 